### PR TITLE
Adding method to EvictingQueue, updating SpanSdk to update if key exists

### LINKED ISSUE #535: Adding duplicate attributes to span
Version: 0.2.0-alpha.197

The `SetAttribute` method name implies that the value for a key is set, implying the key is unique and any additional attempts to set a given key will result in the previous value being overwritten and the last one winning. Currently this is not the case and it behaves like an `AddAttribute` where you are able to add many attributes with the same key.

I feel either this should be renamed to `AddAttribute` - although I'm not sure adding many attributes with the same key makes any sense (unless this is a way of representing arrays maybe?) - or it should behave as the name implies.

### PR DESCRIPTION
**Pull request by eddynaka**
Fixes #535

### CODE REVIEW

**Comment by eddynaka** [comment]
Hi @MikeGoldsmith, can you check if everything is ok? To update the attributes I had to create a new method to find the index and update the value. Is that right?

Thanks.

**Comment by MikeGoldsmith** [comment]
Looks like one of the tests need updating, please can you take a look:
```bash
Microsoft (R) Test Execution Command Line Tool Version 16.3.0
Copyright (c) Microsoft Corporation.  All rights reserved.

Starting test execution, please wait...

A total of 1 test files matched the specified pattern.
[xUnit.net 00:00:02.43]     OpenTelemetry.Trace.Test.SpanTest.BadArguments_SetAttribute_NullKey [FAIL]
  X OpenTelemetry.Trace.Test.SpanTest.BadArguments_SetAttribute_NullKey [3ms]
  Error Message:
   Assert.Equal() Failure
Expected: 4
Actual:   1
  Stack Trace:
     at OpenTelemetry.Trace.Test.SpanTest.BadArguments_SetAttribute_NullKey() in d:\a\1\s\test\OpenTelemetry.Tests\Impl\Trace\SpanTest.cs:line 982
Results File: d:\a\_temp\VssAdministrator_fv-az130_2020-03-31_11_58_13.trx
Test Run Failed.

Total tests: 410
     Passed: 409
     Failed: 1
 Total time: 10.5035 Seconds

```

**Comment by eddynaka** [comment]
@MikeGoldsmith , should I rename the Update to Replace? I will create a new unit test for that method as well. thanks for reviewing.

**Comment by eddynaka** [comment]
@MikeGoldsmith , just updated the test as well and renamed to Replace. I think it's more correct. Let me know if that is ok! Thanks

**Comment by MikeGoldsmith** [comment]
Looks good thanks @eddynaka

File: src/OpenTelemetry/Trace/SpanSdk.cs [hunk 1]
```diff
@@ -333,7 +333,16 @@ public override void SetAttribute(string key, object value)
                         new EvictingQueue<KeyValuePair<string, object>>(this.tracerConfiguration.MaxNumberOfAttributes);
                 }
 
-                this.attributes.Add(new KeyValuePair<string, object>(key ?? string.Empty, sanitizedValue));
+                var attribute = this.attributes.FirstOrDefault(a => a.Key == (key ?? string.Empty));
```
Comment (CodeBlanch) [inline src/OpenTelemetry/Trace/SpanSdk.cs]: @eddynaka Looks like the same block more or less 4 times, should we have a helper method? private void AddOrReplaceAttribute or something?

File: src/OpenTelemetry/Trace/Internal/EvictingQueue.cs [hunk 2]
```diff
@@ -18,6 +18,7 @@
 using System.Collections;
 using System.Collections.Generic;
 using System.Diagnostics;
+using System.Linq;
```
Comment (cijothomas) [inline src/OpenTelemetry/Trace/Internal/EvictingQueue.cs]: Is this namespace used?

File: src/OpenTelemetry/Trace/Internal/EvictingQueue.cs [hunk 3]
```diff
@@ -59,6 +60,16 @@ IEnumerator IEnumerable.GetEnumerator()
             return new Enumerator(this);
         }
 
+        internal void Replace(T item, T newItem)
+        {
+            Debug.Assert(item != null, "Item must not be null");
+            Debug.Assert(newItem != null, "Item must not be null");
+
+            var index = Array.IndexOf(this.items, item);
```
Comment (cijothomas) [inline src/OpenTelemetry/Trace/Internal/EvictingQueue.cs]: some safety check on index? what is item is not found in this.items?

Comment (eddynaka) [reply src/OpenTelemetry/Trace/Internal/EvictingQueue.cs]: Hi @cijothomas , i was validating in the methods that used EvictingQueue.Add method. What do you suggest? Thank you for reviewing

File: src/OpenTelemetry/Trace/SpanSdk.cs (hunk 1 above)
Comment (eddynaka) [reply src/OpenTelemetry/Trace/SpanSdk.cs]: Sure! thanks for the suggestion!

File: src/OpenTelemetry/Trace/Internal/EvictingQueue.cs (hunk 2 above)
Comment (eddynaka) [reply src/OpenTelemetry/Trace/Internal/EvictingQueue.cs]: thanks! i just removed!

File: src/OpenTelemetry/Trace/Internal/EvictingQueue.cs (hunk 3 above)
Comment (eddynaka) [reply src/OpenTelemetry/Trace/Internal/EvictingQueue.cs]: Just checked with this:
```csharp
if (index < 0)
{
   return;
}
```

what do you think?

Comment (cijothomas) [reply src/OpenTelemetry/Trace/Internal/EvictingQueue.cs]: This is better!

Comment (cijothomas) [reply src/OpenTelemetry/Trace/Internal/EvictingQueue.cs]: i'd add a unit test for this as well.

Comment (eddynaka) [reply src/OpenTelemetry/Trace/Internal/EvictingQueue.cs]: Thanks! just added!

**Pull request merged**